File: src/main/javascript/crypto/pbkdf2.js
"use strict";

var nodyn            = require('nodyn');
var Buffer           = require('buffer').Buffer;
var Helper           = org.projectodd.nodyn.buffer.Helper;
var SecretKeyFactory = javax.crypto.SecretKeyFactory;
var PBEKeySpec       = javax.crypto.spec.PBEKeySpec;
var SecretKeySpec    = javax.crypto.spec.SecretKeySpec;

var PBKDF2 = {};

function pbkdf2Sync(password, salt, iterations, keylen) {
  var factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA1");
  var keySpec = new PBEKeySpec( Helper.characters(password), Helper.bytes(salt, 'utf8'), iterations, keylen * 8 );
  var secretKey = factory.generateSecret(keySpec);
  var encoded = secretKey.encoded;
  var keyBuf = new Buffer( encoded );
  return keyBuf;
}

function pbkdf2(password, salt, iterations, keylen, callback) {
  if ( ! callback ) {
    throw new Error( "no callback provided" );
  }
  nodyn.asyncAction( function() {
    return pbkdf2Sync(password, salt, iterations, keylen);
  }, callback );
}

module.exports = {
  pbkdf2Sync: pbkdf2Sync,
  pbkdf2: pbkdf2,
}

File: src/main/java/io/nodyn/crypto/PBKDF2.java
package io.nodyn.crypto;

import io.nodyn.buffer.Buffer;
import java.nio.ByteBuffer;

import javax.crypto.SecretKey;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.PBEKeySpec;
import java.nio.charset.Charset;
import java.security.NoSuchAlgorithmException;
import java.security.spec.InvalidKeySpecException;
import java.security.spec.KeySpec;

/**
 * @author Bob McWhirter
 */
public class PBKDF2 {

    public static ByteBuffer pbkdf2(ByteBuffer password, ByteBuffer salt, int iterations, int keyLen) throws NoSuchAlgorithmException, InvalidKeySpecException {
        SecretKeyFactory factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA1");

        char[] passwordChars = new String(Buffer.extractByteArray(password), Charset.forName("UTF-8")).toCharArray();

        byte[] saltBytes = Buffer.extractByteArray(salt);

        KeySpec keySpec = new PBEKeySpec( passwordChars, saltBytes, iterations, keyLen * 8 );

        SecretKey secretKey = factory.generateSecret(keySpec);

        byte[] keyBytes = secretKey.getEncoded();
        ByteBuffer out = ByteBuffer.allocate(keyBytes.length);
        out.put(keyBytes);
        return out;
    }
}
